utils/chunked_vector.hh

```diff
@@ -430,18 +432,33 @@ chunked_vector<T, max_contiguous_allocation>::make_room(size_t n, bool stop_afte
 }
 
 template <typename T, size_t max_contiguous_allocation>
-void
-chunked_vector<T, max_contiguous_allocation>::do_reserve_for_push_back() {
+typename std::pair<typename chunked_vector<T, max_contiguous_allocation>::chunk_ptr, size_t> chunked_vector<T, max_contiguous_allocation>::get_chunk_before_emplace_back() {
+    size_t new_chunk_capacity;
     if (_capacity == 0) {
         // allocate a bit of room in case utilization will be low
-        reserve(std::clamp(512 / sizeof(T), size_t(1), max_chunk_capacity()));
+        new_chunk_capacity = std::clamp(512 / sizeof(T), size_t(1), max_chunk_capacity());
     } else if (_capacity < max_chunk_capacity() / 2) {
         // exponential increase when only one chunk to reduce copying
-        reserve(_capacity * 2);
+        new_chunk_capacity = _capacity * 2;
     } else {
         // add a chunk at a time later, since no copying will take place
-        reserve((_capacity / max_chunk_capacity() + 1) * max_chunk_capacity());
+        new_chunk_capacity = max_chunk_capacity();
```

Comment:
The new code doesn't call reserve, it allocates a new chunk (which is discardable), that will either replace the first chunk or will be added as a new chunk.  That happens below, in `set_chunk_after_emplace_back`, when the chunk allocated here is applied back to `chunks`, already containing the new items that is pushed or emplaced back.

Reply:
Ok, I understand now. Thanks. It's confusing because I saw both the before and after code side by side, but I guess it's ok and understandable if I look only at the end result.